### PR TITLE
@eessex => mongodb based sessions store

### PR DESCRIPTION
**Pull request by l2succes**
This PR switches the store for editor sessions from in-process memory to a mongodb collection in order to scale across multiple node instances (dynos for heroku).

### CODE REVIEW

File: client/apps/websocket/index.js [hunk 1]
```diff
@@ -8,17 +10,22 @@ const {
...
+  // const filteredSessions = {}
+
+  // for (let [key, session] of Object.entries(sessions)) {
+  //   if (channel.id === session.channel.id) {
+  //     filteredSessions[key] = session
+  //   }
+  // }
+  // return filteredSessions
+
+  sessions.fetch({
```
Comment (l2succes) [inline client/apps/websocket/index.js]: Please ignore: still working on this logic

File: api/apps/sessions/model.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    if (err) {
+      callback(err)
+    }
+
+    const data = {
+      ...omit(input, 'id'),
+      _id: input.id
+    }
+
+    db.sessions.save(data, callback)
```
Comment (l2succes) [inline api/apps/sessions/model.js]: I'm getting the following error here:

```
POST /api/sessions 403 10.307 ms - 104
TypeError: Cannot read property 'save' of undefined
  at /Users/lucsucces/Projects/positron/api/apps/sessions/model.js:84:26
  at internals.Object._validateWithOptions (/Users/lucsucces/Projects/positron/node_modules/joi/lib/types/any/index.js:654:20)
  at module.exports.internals.Any.root.validate (/Users/lucsucces/Projects/positron/node_modules/joi/lib/index.js:121:23)
  at save (/Users/lucsucces/Projects/positron/api/apps/sessions/model.js:75:17)
  at saveSession (/Users/lucsucces/Projects/positron/api/apps/sessions/routes.js:42:17)
  at Layer.handle [as handle_request] (/Users/lucsucces/Projects/positron/node_modules/express/lib/router/layer.js:95:5)
  at next (/Users/lucsucces/Projects/positron/node_modules/express/lib/router/route.js:137:13)
  at Route.dispatch (/Users/lucsucces/Projects/positron/node_modules/express/lib/router/route.js:112:3)
```
Do I need to create collection manually? @kanaabe

File: client/apps/edit/components/edit_container.jsx [hunk 3]
```diff
@@ -48,7 +48,7 @@ export class EditContainer extends Component {
       lastUpdated: null,
       isOtherUserInSession: !!props.currentSession && !isCurrentUserEditing,
       inactivityPeriodEntered: false,
-      shouldShowModal: true,
+      shouldShowModal: !props.article.isNew(),
       sentStopEditingEvent: false
```
Comment (eessex) [inline client/apps/edit/components/edit_container.jsx]: FYI we will be using a plain object in the redux store v soon (after all the redux changes are merged), so it is probably better not to use `article.isNew` and instead check if the id exists

Comment (l2succes) [reply client/apps/edit/components/edit_container.jsx]: Yeah I was thinking about that as well, let's switch it when we replace the backbone model for a plain object

File: api/apps/sessions/routes.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    res.send(session)
+  })
+}
+
+export const deleteSession = (req, res, next) => {
+  destroy(req.params.id, (err) => {
+    if (err) {
+      return next(err)
+    }
+    res.send({id: req.params.id})
```
Comment (eessex) [inline api/apps/sessions/routes.js]: looks like there are test files here but they are empty -- we should definitely test this file!

File: client/apps/articles_list/routes.coffee [hunk 5]
```diff
@@ -32,8 +32,9 @@ Transport = require('lokka-transport-http').Transport
 renderArticles = (res, req, result, published) ->
   res.locals.sd.ARTICLES = result.articles
   channel = res.locals.sd.CURRENT_CHANNEL = req.user?.get('current_channel')
-  res.locals.sd.ARTICLES_IN_SESSION = getSessionsForChannel channel
-  res.locals.sd.HAS_PUBLISHED = published
-  res.render 'index',
-    articles: result.articles || []
-    current_channel: channel
+  getSessionsForChannel channel, (sessions) ->
+    res.locals.sd.ARTICLES_IN_SESSION = sessions
```
Comment (eessex) [inline client/apps/articles_list/routes.coffee]: Is the `getSessionsForChannel` method being called?

File: client/apps/websocket/index.js [hunk 6]
```diff
@@ -8,91 +10,120 @@ const {
...
   const { channel, timestamp, user, article } = data
   const { id, name } = user
-  const currentSession = sessions[article]
-  if (currentSession && currentSession.user.id !== id) {
-    socket.emit(articleLocked, {
-      type: 'ARTICLE_LOCKED',
-      payload: sessions[article]
+
+  console.log(getSessionsForChannel)
+  return getSessionsForChannel(channel).then(sessions => {
```
Comment (eessex) [inline client/apps/websocket/index.js]: stray `console.log`

File: client/apps/websocket/test/index.test.js [hunk 7]
```diff
@@ -1,12 +1,13 @@
...
-  sessions,
   getSessionsForChannel,
   onArticlesRequested,
   onUserCurrentlyEditing,
   onUserStartedEditing,
   onUserStoppedEditing
 } from '../index'
 
+import Sessions from '../../../collections/sessions'
+
```
Comment (eessex) [inline client/apps/websocket/test/index.test.js]: Possible to use an absolute path here?

**Comment by eessex** [comment]
The API side of this should be tested, otherwise just a few minor comments.

File: client/apps/websocket/test/index.test.js (hunk 7 above)
Comment (l2succes) [reply client/apps/websocket/test/index.test.js]: Yes it is, i'll do it

File: client/models/session.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import Backbone from 'backbone'
+import { data as sd } from 'sharify'
+
+export default class Session extends Backbone.Model {
```
Comment (damassi) [inline client/models/session.js]: seeing backbone not written in coffeescript is weirddd

Comment (l2succes) [reply client/models/session.js]: haha true, let's not make a habit of it though :)

**Pull request merged**